DOCS-5190 Create clerk middleware page (#647)
clerk/clerk-docs: docs/references/nextjs/route-protection-requiring-authorization.mdx

---
title: Requiring authorization for route protection | Next.js
description: Learn how to protect the B2B (Business to Business) routes in your Next.js application that require authorization.
---

# Required authorization for route protection

> Whether you're using the App Router, Pages Router, or Middleware, Clerk provides access to a standardized [`auth()`](/docs/references/nextjs/authentication-object) object that you can use to access authorization throughout your application. This allows you to protect routes based on Clerk's [organizations](/docs/organizations/overview) and their roles and permissions.

Author: Most of the guide is using `auth()`. It should be adding `getAuth()` and `useUser()` if anything.

> Whether you're using the App Router, Pages Router, or Middleware, Clerk provides access to a standardized [`auth()`](/docs/references/nextjs/authentication-object) object that you can use to access authorization throughout your application. This allows you to protect routes based on Clerk's [organizations](/docs/organizations/overview) and their roles and permissions.

Reviewer: ```suggestion
Whether you're using the App Router, Pages Router, or Middleware, there are several methods you can use to protect routes within your app such that they require a user to have a certain role or permissions within an [organization](/docs/organizations/overview) to access. This guide will go through a variety of examples of how to do this.
```

You can protect an entire route tree or an individual page by using Clerk helpers based on [permissions](/docs/organizations/verify-user-permissions) or [roles](/docs/organizations/verify-user-permissions#authorize-with-roles). There are two methods that you can use depending on your use case:

> - Using [`auth().protect()`](/docs/references/nextjs/auth#protect) if you want Clerk to redirect user who isn't signed in to the sign-in route, or return a 404 if the user does not have the role or permission

Reviewer 2: ```suggestion
- Using [`auth().protect()`](/docs/references/nextjs/auth#protect) if you want Clerk to redirect a user who isn't signed in to the sign-in route, or return a 404 if the user does not have the role or permission
```

You can use protect queries and mutations in your Route Handlers or Server Actions in two ways:

> - `auth().protect()` if you want Clerk to redirect user who isn't signed in to the sign-in route, or return a 404 if the user does not have the role or permission

Reviewer 2: ```suggestion
- `auth().protect()` if you want Clerk to redirect a user who isn't signed in to the sign-in route, or return a 404 if the user does not have the role or permission
```

- `auth().has()` if you want to customize the behavior or error message

```tsx filename="/app/api/dashboard/route.ts"
import { auth } from '@clerk/nextjs/server';

export async function GET() {
  if (!auth().has({ permission: 'org:sys_memberships:manage' })) {
    return Response.json(
      { error: 'Error: User does not have required permissions' },
      { status: 401 },
    );
  }

  // Add your Route Handler logic here

  return Response.json({ userId: auth().userId });
}
```

## Pages Router

### Protect a page

> `getAuth()` does not provide access to the `has()` method. We recommend that you handle authorization for routes in the Middleware, where `clerkMiddleware()` will provide access to `auth().protect()` and `auth().has()`.

Reviewer 2: wait, I thought has is available on the Auth object - so why doesn't getAuth provide access to has?

and the example before was using useAuth - so why did we remove it?
